source/material_model/viscoelastic.cc

```diff
@@ -72,26 +65,69 @@ namespace aspect
           for (unsigned int c=0; c<in.composition[i].size(); ++c)
             out.reaction_terms[i][c] = 0.0;
 
-          // Average viscosity and shear modulus
+          // Average the viscous viscosity and the shear modulus over the compositions
           const double average_viscosity = MaterialUtilities::average_value(volume_fractions, viscosities, viscosity_averaging);
           average_elastic_shear_moduli[i] = MaterialUtilities::average_value(volume_fractions, elastic_shear_moduli, viscosity_averaging);
 
-          // Average viscoelastic (e.g., effective) viscosity (equation 28 in Moresi et al., 2003, J. Comp. Phys.)
+          // Average viscoelastic (e.g., effective) viscosity (equation 28 in Moresi et al., 2003, J. Comp. Phys.).
+          // This viscosity is scaled with the timestep ratio $\frac{\Delta t_c}{\Delta t_{el}}$ in the
+          // calculate_viscoelastic_viscosity function.
           out.viscosities[i] = elastic_rheology.calculate_viscoelastic_viscosity(average_viscosity,
                                                                                  average_elastic_shear_moduli[i]);
 
           // Fill the material properties that are part of the elastic additional outputs
-          if (ElasticAdditionalOutputs<dim> *elastic_out = out.template get_additional_output<ElasticAdditionalOutputs<dim>>())
+          if (ElasticAdditionalOutputs<dim> *elastic_additional_out = out.template get_additional_output<ElasticAdditionalOutputs<dim>>())
             {
-              elastic_out->elastic_shear_moduli[i] = average_elastic_shear_moduli[i];
+              elastic_additional_out->elastic_shear_moduli[i] = average_elastic_shear_moduli[i];
             }
         }
 
+      // Fill the body force term, viscoelastic strain rate and viscous dissipation.
       elastic_rheology.fill_elastic_outputs(in, average_elastic_shear_moduli, out);
+      // Fill the reaction terms to apply the rotation of the stresses into the current timestep.
       elastic_rheology.fill_reaction_outputs(in, average_elastic_shear_moduli, out);
+      // Fill the reaction_rates that during operator splitting apply the stress update of the previous
+      // timestep to the advected and rotated stress computed in the previous timestep ($\tau^{0adv}$)
+      // to obtain $\tau^{t}$.
+      if (this->get_parameters().use_operator_splitting ||
+          ((this->get_parameters().mapped_particle_properties).count(this->introspection().compositional_index_for_name("ve_stress_xx"))))
```

Comment:
you can use the `introspection` to ask for material type here (see my comment in visco_plastic).

Reply:
same here, I've removed this check.